File: apps/api-gql/internal/delivery/gql/mappers/overlays-tts.go
package mappers

import (
	"github.com/twirapp/twir/apps/api-gql/internal/delivery/gql/gqlmodel"
	"github.com/twirapp/twir/apps/api-gql/internal/entity"
)

func MapTTSOverlayEntityToGQL(e entity.TTSOverlay) gqlmodel.TTSOverlay {
	return gqlmodel.TTSOverlay{
		ID:                                 e.ID,
		Enabled:                            e.Settings.Enabled,
		Voice:                              e.Settings.Voice,
		DisallowedVoices:                   e.Settings.DisallowedVoices,
		Pitch:                              int(e.Settings.Pitch),
		Rate:                               int(e.Settings.Rate),
		Volume:                             int(e.Settings.Volume),
		DoNotReadTwitchEmotes:              e.Settings.DoNotReadTwitchEmotes,
		DoNotReadEmoji:                     e.Settings.DoNotReadEmoji,
		DoNotReadLinks:                     e.Settings.DoNotReadLinks,
		AllowUsersChooseVoiceInMainCommand: e.Settings.AllowUsersChooseVoiceInMainCommand,
		MaxSymbols:                         int(e.Settings.MaxSymbols),
		ReadChatMessages:                   e.Settings.ReadChatMessages,
		ReadChatMessagesNicknames:          e.Settings.ReadChatMessagesNicknames,
		CreatedAt:                          e.CreatedAt,
		UpdatedAt:                          e.UpdatedAt,
		ChannelID:                          e.ChannelID,
	}
}

func TTSUserSettingTo(e entity.TTSUserSettings) gqlmodel.TTSUserSettings {
	return gqlmodel.TTSUserSettings{
		UserID:         e.UserID,
		Rate:           e.Rate,
		Pitch:          e.Pitch,
		Voice:          e.Voice,
		IsChannelOwner: e.IsChannelOwner,
	}
}

File: frontend/dashboard/src/api/overlays-tts.ts
import { useQuery } from '@urql/vue'
import { createGlobalState } from '@vueuse/core'

import type { OverlaysTtsQuery } from '@/gql/graphql.js'

import { useMutation } from '@/composables/use-mutation.js'
import { graphql } from '@/gql/gql.js'

export type TTSOverlay = Omit<OverlaysTtsQuery['overlaysTTS'], '__typename' | 'channel'>

const invalidationKey = 'TTSOverlayInvalidateKey'

export const useTTSOverlayApi = createGlobalState(() => {
	const useQueryTTS = () =>
		useQuery({
			variables: {},
			context: { additionalTypenames: [invalidationKey] },
			query: graphql(`
				query OverlaysTTS {
					overlaysTTS {
						id
						enabled
						voice
						disallowedVoices
						pitch
						rate
						volume
						doNotReadTwitchEmotes
						doNotReadEmoji
						doNotReadLinks
						allowUsersChooseVoiceInMainCommand
						maxSymbols
						readChatMessages
						readChatMessagesNicknames
						createdAt
						updatedAt
						channelId
					}
				}
			`),
		})

	const useMutationUpdateTTS = () =>
		useMutation(
			graphql(`
				mutation OverlaysTTSUpdate($input: TTSUpdateInput!) {
					overlaysTTSUpdate(input: $input) {
						id
						enabled
						voice
						disallowedVoices
						pitch
						rate
						volume
						doNotReadTwitchEmotes
						doNotReadEmoji
						doNotReadLinks
						allowUsersChooseVoiceInMainCommand
						maxSymbols
						readChatMessages
						readChatMessagesNicknames
						createdAt
						updatedAt
						channelId
					}
				}
			`),
			[invalidationKey]
		)

	const useQueryTTSUsersSettings = () =>
		useQuery({
			variables: {},
			query: graphql(`
				query OverlaysTTSUsersSettings {
					overlaysTTSUsersSettings {
						userId
						twitchProfile {
							id
							login
							displayName
							profileImageUrl
						}
						rate
						pitch
						voice
						isChannelOwner
					}
				}
			`),
		})

	const useMutationDeleteTTSUsersSettings = () =>
		useMutation(
			graphql(`
				mutation OverlaysTTSUsersDelete($userIds: [String!]!) {
					overlaysTTSUsersDelete(userIds: $userIds)
				}
			`),
			[]
		)

	const useQueryTTSGetInfo = () =>
		useQuery({
			variables: {},
			query: graphql(`
				query OverlaysTTSGetInfo {
					overlaysTTSGetInfo {
						voicesInfo {
							key
							info {
								country
								gender
								lang
								name
								no
							}
						}
					}
				}
			`),
		})

	return {
		useQueryTTS,
		useMutationUpdateTTS,
		useQueryTTSUsersSettings,
		useMutationDeleteTTSUsersSettings,
		useQueryTTSGetInfo,
	}
})
